protocol_decoder/src/decoding.rs

```diff
@@ -281,7 +293,45 @@ impl ProcessedBlockTrace {
                 let dummy_txn = create_dummy_gen_input(other_data, initial_trie_state, 0);
                 gen_inputs.insert(0, dummy_txn);
             }
-            _ => (),
+            _ => dummys_added = false,
+        }
+
+        dummys_added
+    }
+
+    /// The withdrawls are always represented as a single "dummy" txn at the end
+    /// of the block. However, if no dummies have already been added, then
+    /// we need to append one to the end. If dummies have been added, then
+    /// add it to the last one.
+    fn add_withdrawls_to_txns(
+        txn_ir: &mut Vec<TxnProofGenIR>,
+        other_data: &OtherBlockData,
+        final_trie_state: &PartialTrieState,
+        withdrawals: Vec<(Address, U256)>,
+        dummies_already_added: bool,
+    ) {
+        if withdrawals.is_empty() {
```

Comment:
nit: I'd rather have this condition for calling `add_withdrawls_to_txns`, not inside, but up to you